rfc/src/rfcs/0011-list.md

```diff
@@ -0,0 +1,157 @@
+    attr_impl::requires(attr, item)
+}
+```
+
+The `item` `TokenStream` gives us the function declaration to which the attribute is applied, e.g. `fn foo() {}`, but that is insufficient to uniquely identify the function or construct the module tree. Instead, we will change the macro expansion handler (`ContractConditionsHandler`) to insert the contracts text as a separate attribute, e.g.:
+
+```rust
+// these are wrappers we already generate for contracts
+#[kanitool::recursion_check = #recursion_name]
+#[kanitool::checked_with = #check_name]
+...
+#[kanitool::contract_text(requires(x == 1))] // this is new
+```
+
+Then, in `kani-compiler`, we add a new branch of the code generation process for the `list` subcommand. When `codegen_crate` encounters this reachability type, it will:
+- Construct a `CodegenUnits` object.
+- Call a new `CodegenUnits` method that filters local crate instances to those with the `contract_text` attribute, then stores the following in a new `contracted_functions` field:
+    - the type of contract (`Requires`, `Ensures`, or `Modifies`), 
+    - the text of the contract, and 
+    - the fully qualified path to it.
+- Call `write_metadata` on the `CodegenUnits` object, which will construct a `KaniMetadata` object with a new field containing the contracted function information, then write it to a file.
+
+### Constructing the Tree
+
+Once we've made the above updates to `KaniMetadata`, that struct will have all of the information we need for `kani list`. 
+
+To print the tree, Kani will read the metadata file and construct an intermediate data structure that maps each module to a list of its metadata. (This list will be sorted so that the output is stable across runs). Then, we'll iterate through that structure to print the tree. If the user passes a `--filter` option, Kani will only construct the map for the relevant metadata--e.g., if the user passes `--filter harnesses`, we will only map each module to its harnesses and ignore the collected contracts.
+
+Kani will parse the module paths to print the tree, inserting a new level for each delimiter `::`. For example, `example::implementation::verify` would parse to:
+
+```
+crate example
+├── mod implementation
+│   └── mod verify
+```
+
+### Calculating Totals
+Since we already have the information about each harness, contract, etc., calculating totals is as simple as taking the length of those data structures.
+
+## Rationale and alternatives
+
+We organize the output by module so that users can compare Kani integration across their package. The output can be quite long, since Kani prints the whole module tree. As an alternative, we could support a more compact output format, e.g. one that just maps the Rust path (`example::verify::foo`) to the metadata. This format would have the advantage of being more concise, but the disadvantage of making it harder for the user to visualize how Kani fits into their package structure.
```

Comment:
I actually prefer the more compact view. It's simpler and concise. Since we are sorting by the function name, it will naturally sort them by their modules.

For the json output, we could add a field `module` and I would also suggest adding the file location.